File: src/main/java/eu/mshade/enderframe/entity/Player.java
package eu.mshade.enderframe.entity;

import eu.mshade.enderframe.Agent;
import eu.mshade.enderframe.GameMode;
import eu.mshade.enderframe.inventory.Inventory;
import eu.mshade.enderframe.inventory.InventoryClickStore;
import eu.mshade.enderframe.inventory.NamedInventory;
import eu.mshade.enderframe.inventory.PlayerInventory;
import eu.mshade.enderframe.mojang.GameProfile;
import eu.mshade.enderframe.protocol.MinecraftProtocolVersion;
import eu.mshade.enderframe.protocol.MinecraftSession;
import eu.mshade.enderframe.scoreboard.Scoreboard;
import eu.mshade.enderframe.world.World;
import eu.mshade.enderframe.world.chunk.Chunk;
import eu.mshade.enderframe.world.Location;
import eu.mshade.enderframe.world.Vector;
import eu.mshade.enderframe.world.border.WorldBorder;

import java.net.InetSocketAddress;
import java.util.Collection;
import java.util.Queue;
import java.util.UUID;
import java.util.concurrent.ConcurrentLinkedQueue;

public abstract class Player extends Entity implements Agent {

    protected InetSocketAddress inetSocketAddress;
    protected MinecraftProtocolVersion minecraftProtocolVersion;
    protected int ping;
    protected String displayName;
    protected GameMode gameMode = GameMode.UNKNOWN;
    protected GameProfile gameProfile;
    protected boolean invulnerable;
    protected boolean flying;
    protected boolean allowFlying;
    protected boolean instantBreak;
    protected float flyingSpeed;
    protected float walkSpeed;
    protected boolean sneaking;
    protected PlayerInventory playerInventory;
    protected Inventory openedInventory;
    protected Queue<Chunk> lookAtChunks = new ConcurrentLinkedQueue<>();
    protected InventoryClickStore inventoryClickStore = new InventoryClickStore();
    protected Queue<Scoreboard> lookAtScoreboard = new ConcurrentLinkedQueue<>();
    protected Queue<WorldBorder> lookAtWorldBorders = new ConcurrentLinkedQueue<>();

    public Player(Location location, Vector velocity, int entityId, UUID uuid, EntityKey entityKey) {
        super(location, velocity, entityId, uuid, entityKey);
    }

    public Player(Location location, int entityId, EntityKey entityKey) {
        super(location, entityId, entityKey);
    }

    public Player(Location location, Vector velocity, int entityId, UUID uuid) {
        super(location, velocity, entityId, uuid, EntityType.INSTANCE.getPLAYER());
    }

    public Player(Location location, int entityId) {
        super(location, entityId, EntityType.INSTANCE.getPLAYER());
    }

    public String getName(){
        return getGameProfile().getName();
    }

    public InetSocketAddress getInetSocketAddress(){
        return this.inetSocketAddress;
    }

    public void setInetSocketAddress(InetSocketAddress inetSocketAddress){
        this.inetSocketAddress = inetSocketAddress;
    }

    public void setMinecraftProtocolVersion(MinecraftProtocolVersion minecraftProtocolVersion) {
        this.minecraftProtocolVersion = minecraftProtocolVersion;
    }

    public MinecraftProtocolVersion getMinecraftProtocolVersion(){
        return this.minecraftProtocolVersion;
    }

    public int getPing(){
        return this.ping;
    }

    public void setPing(int ping){
        this.ping = ping;
    }

    public GameMode getGameMode(){
        return this.gameMode;
    }

    public void setGameMode(GameMode gameMode){
        this.gameMode = gameMode;
    }

    public GameProfile getGameProfile(){
        return this.gameProfile;
    }

    public void setGameProfile(GameProfile gameProfile){
        this.gameProfile = gameProfile;
    }

    public String getDisplayName(){
        return this.displayName;
    }

    public void setDisplayName(String name){
        this.displayName = name;
    }

    public boolean isInvulnerable() {
        return invulnerable;
    }

    public void setInvulnerable(boolean invulnerable) {
        this.invulnerable = invulnerable;
    }

    public boolean isFlying() {
        return flying;
    }

    public void setFlying(boolean flying) {
        this.flying = flying;
    }

    public boolean isAllowFlying() {
        return allowFlying;
    }

    public void setAllowFlying(boolean allowFlying) {
        this.allowFlying = allowFlying;
    }

    public boolean isInstantBreak() {
        return instantBreak;
    }

    public void setInstantBreak(boolean instantBreak) {
        this.instantBreak = instantBreak;
    }

    public float getFlyingSpeed() {
        return flyingSpeed;
    }

    public void setFlyingSpeed(float flyingSpeed) {
        this.flyingSpeed = flyingSpeed;
    }

    public float getWalkSpeed() {
        return walkSpeed;
    }

    public void setWalkSpeed(float walkSpeed) {
        this.walkSpeed = walkSpeed;
    }

    public boolean isSneaking() {
        return sneaking;
    }

    public void setSneaking(boolean sneaking) {
        this.sneaking = sneaking;
    }

    public Collection<Chunk> getLookAtChunks(){
        return this.lookAtChunks;
    }

    public boolean hasLookAtChunk(Chunk chunk){
        return this.lookAtChunks.contains(chunk);
    }

    public PlayerInventory getInventory() {
        return playerInventory;
    }

    public Inventory getOpenedInventory() {
        return openedInventory;
    }

    public void setOpenedInventory(Inventory openedInventory) {
        this.openedInventory = openedInventory;
    }

    public void openInventory(NamedInventory inventory){
        setOpenedInventory(inventory);
        getMinecraftSession().sendOpenInventory(inventory);
        getMinecraftSession().sendItemStacks(inventory);
        inventory.addWatcher(this);
    }



    public InventoryClickStore getInventoryBufferStore() {
        return inventoryClickStore;
    }

    public Collection<Scoreboard> getLookAtScoreboard() {
        return this.lookAtScoreboard;
    }

    public Collection<WorldBorder> getLookAtWorldBorders(){
        return this.lookAtWorldBorders;
    }

    public boolean hasLookAtWorldBorder(WorldBorder worldBorder){
        return this.lookAtWorldBorders.contains(worldBorder);
    }


    public abstract void joinWorld(World world);

    public abstract MinecraftSession getMinecraftSession();

}

File: src/main/java/eu/mshade/enderframe/entity/Player.kt
package eu.mshade.enderframe.entity

import eu.mshade.enderframe.Agent
import eu.mshade.enderframe.GameMode
import eu.mshade.enderframe.entity.metadata.EntityMetadataKey
import eu.mshade.enderframe.inventory.*
import eu.mshade.enderframe.item.ItemStack
import eu.mshade.enderframe.mojang.GameProfile
import eu.mshade.enderframe.protocol.MinecraftProtocolVersion
import eu.mshade.enderframe.protocol.MinecraftSession
import eu.mshade.enderframe.scoreboard.Scoreboard
import eu.mshade.enderframe.world.Location
import eu.mshade.enderframe.world.Vector
import eu.mshade.enderframe.world.World
import eu.mshade.enderframe.world.border.WorldBorder
import eu.mshade.enderframe.world.chunk.Chunk
import java.net.InetSocketAddress
import java.util.*
import java.util.concurrent.ConcurrentLinkedQueue

abstract class Player : Entity, Agent, Equipable {

    var inetSocketAddress: InetSocketAddress? = null
    var minecraftProtocolVersion: MinecraftProtocolVersion? = null
    var ping = 0
    var displayName: String? = null
    var gameMode = GameMode.UNKNOWN
    var gameProfile: GameProfile? = null
        protected set
    var isInvulnerable = false
    var isFlying = false
    var isAllowFlying = false
    var isInstantBreak = false
    var flyingSpeed = 0f
    var walkSpeed = 0f
    var inventory: PlayerInventory? = null
        protected set
    var openedInventory: Inventory? = null
    val lookAtChunks: Queue<Chunk> = ConcurrentLinkedQueue()
    var inventoryBufferStore = InventoryClickStore()
        protected set
    private var lookAtScoreboard: Queue<Scoreboard> = ConcurrentLinkedQueue()
    private var lookAtEntity: Queue<Entity> = ConcurrentLinkedQueue()
    private var worldBorder: WorldBorder? = null

    constructor(location: Location, velocity: Vector, entityId: Int, uuid: UUID, entityKey: EntityKey) : super(location, velocity, entityId, uuid, entityKey)
    constructor(location: Location, entityId: Int, entityKey: EntityKey) : super(location, entityId, entityKey)
    constructor(location: Location, velocity: Vector, entityId: Int, uuid: UUID) : super(location, velocity, entityId, uuid, EntityType.PLAYER)
    constructor(location: Location, entityId: Int) : super(location, entityId, EntityType.PLAYER)

    val name: String?
        get() = gameProfile?.name


    fun hasLookAtChunk(chunk: Chunk): Boolean {
        return lookAtChunks.contains(chunk)
    }

    fun hasLookAtChunk(chunkX: Int, chunkZ: Int): Boolean {
        return lookAtChunks.stream().anyMatch { chunk: Chunk -> chunk.x == chunkX && chunk.z == chunkZ }
    }

    fun openInventory(inventory: NamedInventory) {
        openedInventory = inventory
        minecraftSession.sendOpenInventory(inventory)
        minecraftSession.sendItemStacks(inventory)
        inventory.addWatcher(this)
    }

    fun getLookAtScoreboard(): MutableCollection<Scoreboard> {
        return lookAtScoreboard
    }

    fun getWorldBorder(): WorldBorder? {
        return worldBorder
    }

    fun setWorldBorder(worldBorder: WorldBorder) {
        this.worldBorder = worldBorder
        if (!worldBorder.isWatching(this)) {
            worldBorder.addWatcher(this)
        }
    }

    abstract fun joinWorld(world: World?)

    abstract val minecraftSession: MinecraftSession

    fun sendChunk(chunk: Chunk){
        if (chunk.isWatching(this)) return
        minecraftSession.sendChunk(chunk)
        chunk.addWatcher(this)
        this.lookAtChunks.add(chunk)
    }

    fun isSneaking(): Boolean {
        var metadataKeyValue = metadata.getMetadataKeyValue(EntityMetadataKey.CROUCHED)
        return metadataKeyValue != null && metadataKeyValue.metadataValue as Boolean
    }

    fun getLookAtEntity(): Collection<Entity> {
        return lookAtEntity
    }

    fun addLookAtEntity(entity: Entity) {
        if (lookAtEntity.contains(entity)) return
        lookAtEntity.add(entity)
    }

    fun removeLookAtEntity(entity: Entity) {
        lookAtEntity.remove(entity)
    }

    override fun getEquipment(slot: EquipmentSlot): ItemStack? {
        return inventory?.getItemStack(slot)
    }

    override fun setEquipment(slot: EquipmentSlot, itemStack: ItemStack?) {
        inventory?.setItemStack(slot, itemStack)
    }

    override fun getEquipments(): Map<EquipmentSlot, ItemStack> {
        val equipments = mutableMapOf<EquipmentSlot, ItemStack>()

        for (slot in EquipmentSlot.values()) {
            getEquipment(slot)?.let { equipments[slot] = it }
        }

        return equipments
    }
}
